ecosystem/python/sdk/aptos_sdk/account_sequence_number.py

```diff
@@ -0,0 +1,178 @@
+    Aptos faucet:
+    * Submit up to 50 transactions per account in parallel with a timeout of 20 seconds
+    * If local assumes 50 are in flight, determine the actual committed state from the network
+    * If there are less than 50 due to some being committed, adjust the window
+    * If 50 are in flight Wait .1 seconds before re-evaluating
+    * If ever waiting more than 30 seconds restart the sequence number to the current on-chain state
+    Assumptions:
+    * Accounts are expected to be managed by a single AccountSequenceNumber and not used otherwise.
+    * They are initialized to the current on-chain state, so if there are already transactions in
+      flight, they make take some time to reset.
+    * Accounts are automatically initialized if not explicitly
+
+    Notes:
+    * This is co-routine safe, that is many async tasks can be reading from this concurrently.
+    * The synchronize method will create a barrier that prevents additional next_sequence_number
+      calls until it is complete.
+    * This only manages the distribution of sequence numbers it does not help handle transaction
+      failures.
+    """
+
+    client: RestClient
+    account: AccountAddress
+    lock = asyncio.Lock
+
+    maximum_in_flight: int = 100
+    maximum_wait_time = 30
+    sleep_time = 0.01
+
+    last_committed_number: Optional[int]
+    current_number: Optional[int]
+
+    def __init__(self, client: RestClient, account: AccountAddress):
+        self.client = client
+        self.account = account
+        self.lock = asyncio.Lock()
+
+        self.last_uncommitted_number = None
+        self.current_number = None
+
+    async def next_sequence_number(self, block: bool = True) -> Optional[int]:
+        """
+        Returns the next sequence number available on this account. This leverages a lock to
+        guarantee first-in, first-out ordering of requests.
+        """
+        await self.lock.acquire()
+        try:
+            if self.last_uncommitted_number is None or self.current_number is None:
+                await self.initialize()
+            if (
+                self.current_number - self.last_uncommitted_number
+                >= self.maximum_in_flight
+            ):
+                await self.__update()
+                start_time = time.time()
+                while (
+                    self.current_number - self.last_uncommitted_number
+                    >= self.maximum_in_flight
+                ):
+                    if not block:
+                        return None
+                    await asyncio.sleep(self.sleep_time)
+                    if time.time() - start_time > self.maximum_wait_time:
+                        logging.warn(
+                            f"Waited over 30 seconds for a transaction to commit, resyncing {self.account.address().hex()}"
+                        )
+                        await self.initialize()
+                    else:
+                        await self.__update()
```

Comment:
Personally, I prefer comments to tell what we're checking for each of these.

It looks like:
1. Check if the queue is initialized
2. Check if there is over the number in flight
3. Sleep if it's blocking on transactions, otherwise continue to send transactions (need to define `block` in docs)
4. Poll and block to see if the mempool has cleared any transactions
5. If it's clear (or block is false), submit the transaction, and let the next one grab the lock.